infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariablesSorter.java

```diff
@@ -0,0 +1,202 @@
+    if (aTransRefs.contains(b.getName())) {
+      // a depends on b, hence a > b.
+      return 1;
+    } else if (bTransRefs.contains(a.getName())) {
+      // b depends on a -> a < b.
+      return -1;
+    }
+
+    // ok, the two env vars don't depend on each other, so we actually don't care too much about
+    // the order. We need to be consistent and total, though...
+
+    int sizeDiff = aTransRefs.size() - bTransRefs.size();
+    if (sizeDiff != 0) {
+      return sizeDiff;
+    }
+
+    // ah, the two env vars have the same number of references, so we need to take a "slow path"
+    // in determining the total order.
+    Iterator<String> aIt = aRefs.iterator();
+    Iterator<String> bIt = bRefs.iterator();
+    while (aIt.hasNext()) {
+      if (!bIt.hasNext()) {
+        // a has more references than b
+        return 1;
+      }
+
+      String aRef = aIt.next();
+      String bRef = bIt.next();
+
+      int res = aRef.compareTo(bRef);
+      if (res != 0) {
+        return res;
+      }
+    }
+
+    return 0;
+  }
+
+  private SortedSet<String> transitiveClosure(String variableName) {
+    SortedSet<String> recRefs = recursiveReferenceMap.get(variableName);
+    if (recRefs == null) {
+      recRefs = new TreeSet<>(naturalOrder());
+      fillRecursiveReferences(variableName, referenceMap, recRefs);
+      recursiveReferenceMap.put(variableName, recRefs);
+    }
+    return recRefs;
+  }
+
+  @VisibleForTesting
+  static SortedSet<String> findReferences(EnvVar var) throws CycleException {
+    String val = var.getValue();
+
+    Matcher matcher = REFERENCE_PATTERN.matcher(val);
+
+    SortedSet<String> ret = new TreeSet<>(naturalOrder());
+
+    while (matcher.find()) {
+      int start = matcher.start();
+
+      // the variable reference can be escaped using a double $, e.g. $$(VAR) is not a reference
+      if (start > 0 && val.charAt(start - 1) == '$') {
+        continue;
+      }
+
+      // extract the variable name out of the reference $(NAME) -> NAME
+      String refName = matcher.group().substring(2, matcher.group().length() - 1);
+      if (refName.equals(var.getName())) {
+        throw new CycleException(refName);
+      }
+      ret.add(refName);
+    }
+
+    return ret;
+  }
+
+  private static void fillRecursiveReferences(
+      String name, Map<String, SortedSet<String>> directReferences, SortedSet<String> result)
+      throws CycleException {
+
+    Set<String> nameRefs = directReferences.get(name);
+    if (nameRefs == null || nameRefs.isEmpty()) {
+      return;
+    }
+
+    if (result.contains(name)) {
+      throw new CycleException(name);
+    }
+
+    for (String refName : nameRefs) {
+      result.add(refName);
+      fillRecursiveReferences(refName, directReferences, result);
+    }
+  }
+
+  static final class CycleException extends RuntimeException {
+
+    CycleException(String variableName) {
+      super(format("Recursive definition of environment variable %s", variableName));
```

Comment:
Can you elaborate on whether it is this a user-facing error?

Reply:
The message is reused  in the `InfrastructureException` thrown from `sort`